File: src/main/kotlin/xyz/mahmoudahmed/dsl/BranchTargetBuilder.java
package xyz.mahmoudahmed.dsl;

import java.util.function.Function;

public class BranchTargetBuilder {
    private final BranchActionBuilder parent;
    private final String sourcePath;
    private final String targetPath;
    private final Function<Object, Object> transformation;

    BranchTargetBuilder(BranchActionBuilder parent, String sourcePath, String targetPath,
                        Function<Object, Object> transformation) {
        this.parent = parent;
        this.sourcePath = sourcePath;
        this.targetPath = targetPath;
        this.transformation = transformation;
    }

    public BranchActionBuilder end() {
        parent.addAction(new SimpleMapping(sourcePath, targetPath, transformation, null));
        return parent;
    }
}

File: src/main/kotlin/xyz/mahmoudahmed/dsl/conditional/BranchTargetBuilder.kt
package xyz.mahmoudahmed.dsl.conditional

import xyz.mahmoudahmed.dsl.core.SimpleMapping

/**
 * Builder for finalizing a mapping within a conditional branch.
 */
class BranchTargetBuilder(
    private val parent: BranchActionBuilder,
    private val sourcePath: String,
    private val targetPath: String,
    private val transformation: ((Any) -> Any)?
) {
    /**
     * Finishes the mapping configuration.
     *
     * @return The parent branch action builder
     */
    fun end(): BranchActionBuilder {
        val rule = SimpleMapping(sourcePath, targetPath, transformation, null)
        parent.addAction(rule)
        return parent
    }
}